plugin/src/test/java/hudson/plugins/swarm/SwarmClientIntegrationTest.java

```diff
@@ -148,6 +163,83 @@ public void addRemoveLabelsViaFileWithoutUniqueIdLong() throws Exception {
+                j,
+                processDestroyer,
+                temporaryFolder,
+                "-pidFile",
+                getPidFile().getAbsolutePath());
+
+        // Try to start a second and ensure it fails to run.
+        // Do not wait for it to come up on the server side.
+        TestUtils.SwarmClientProcessWrapper node2 = TestUtils.runSwarmClient(
+                "agent_fail",
+                j,
+                processDestroyer,
+                temporaryFolder,
+                "-pidFile",
+                getPidFile().getAbsolutePath());
+
+        node2.process.waitFor(5, TimeUnit.SECONDS);
+        assertFalse("Second client should fail to start", node2.process.isAlive());
+        assertEquals("Exit code", 1, node2.process.exitValue());
+        assertTrue("Log message mentions 'already exists'",
+                Files.readAllLines(node2.stderr.toPath()).stream().anyMatch(line -> line.contains("already exists")));
+    }
+
+    @Test
+    public void pidFileForStaleProcessIsIgnored() throws Exception {
+
+        Files.write(getPidFile().toPath(), "66000".getBytes());
+
+        // Pid file should be ignored since the process isn't running.
+        TestUtils.SwarmClientProcessWrapper node = TestUtils.runSwarmClient(
+                "agent0",
+                j,
+                processDestroyer,
+                temporaryFolder,
+                "-pidFile",
+                getPidFile().getAbsolutePath());
+
+        // ensure that the client starts.
+        TestUtils.waitForNode("agent0", j);
+
+        int newPid = NumberUtils.toInt(
+                new String(Files.readAllBytes(getPidFile().toPath()), UTF_8));
+
+        // Java Process doesn't provide the pid, so we have to work around it.
+        // Find all of our child processes, one of them must be the client we just started,
+        // and thus would match the pid in the pidFile.
+        OSProcess[] childProcesses = os.getChildProcesses(os.getProcessId(), 0, null);
+        assertTrue("Pid in pidFile must match our new pid",
+                Arrays.stream(childProcesses).anyMatch(proc -> proc.getProcessID() == newPid));
+    }
+
+    @Test
+    public void pidFileDeletedOnExit() throws Exception {
+        TestUtils.SwarmClientProcessWrapper node = TestUtils.runSwarmClient(
+                "agentDeletePid",
+                j,
+                processDestroyer,
+                temporaryFolder,
+                "-pidFile",
+                getPidFile().getAbsolutePath());
+        Thread.sleep(1000); //ensure the process writes the pid
+        assertTrue("Pid file created", getPidFile().exists());
+        node.process.destroy();
+        node.process.waitFor(5, TimeUnit.SECONDS);
+        assertFalse("Client should exit on kill", node.process.isAlive());
+        assertFalse("Pid file removed", getPidFile().exists());
+    }
+
+    public File getPidFile() {
+        return new File(temporaryFolder.getRoot(), "__tmp.pid.file.swarm__");
```

Comment:
I can make this change, but `File.createTempFile()` isn't exactly the same thing since it actually creates the file.  It's probably not a dramatic difference, but in the tests we'd then be testing the "ignore invalid pid" rather than "pid file correctly written when not present".  I can do `createTempFile` then delete it, of course, but that's kind of ugly, which is why I wrote it the way I did.

Reply:
Actually creating and then deleting seems fine in the test.